File: MyCPLHiQ.py
class Stack():
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def push(self, item):
        return self.items.append(item)

    def pop(self):
        return self.items.pop()

    def getElements(self):
        return self.items


def findmyrow(board, emp_hole):
    count = 0
    for x in range(0, 5):
        for y in range(0, x + 1):
            if (count == init_hole):
                board[x][y] = 0
            else:
                board[x][y] = 1
            count = count + 1
    return board


def moves(x, y):
    c = 0
    for i in range(0, 5):
        for j in range(0, i + 1):
            if (i == x and j == y):
                return c
            c = c + 1


def check_possiblemove(board):
    pegmove = []
    for x in range(0, 5):
        for y in range(0, x + 1):
            if (board[x][y] == 0):
                pegmove.append([x, y])
    return pegmove


def mypegs(board, x, y):
    if (board[x][y] == 0):
        return 0
    else:
        return 1


def uploadmymoves(board):
    l = check_possiblemove(board)
    bettersol = []
    for x in range(0, len(l)):
        moves_allowed = l[x]
        for y in range(-2, 4, 2):
            for z in range(-2, 4, 2):
                if (y == 0 and z == 0):
                    continue
                var1 = 0
                var2 = 0
                var1 = moves_allowed[0] + y
                var2 = moves_allowed[1] + z
                if (y == 2 and z == -2): continue
                if (y == -2 and z == 2): continue
                if (var1 >= 0 and var2 >= 0 and var1 < 5 and var2 < 5 and var1 >= var2 and mypegs(board, var1,
                                                                                                       var2) == 1):
                    var1 = int((var1 + moves_allowed[0]) / 2)
                    var2 = int((var2 + moves_allowed[1]) / 2)
                    if (mypegs(board, var1, var2) == 1):
                        bettersol.append([moves(moves_allowed[0] + y, moves_allowed[1] + z), moves(var1, var2),
                                          moves(moves_allowed[0], moves_allowed[1])])
    return bettersol


def jump(board, num, poss_move):
    c = 0
    for i in range(0, 5):
        for j in range(0, i + 1):
            if (c == num):
                board[i][j] = poss_move
            c = c + 1


def checkforsolution(board, checkhole):
    global peg
    if (checkhole == peg):
        if (checkhole == 13):
            print("This is the best possible solution")
            pegmove = stack.getElements()
            z = 0
            for i in range(0, 13):
                print( pegmove[i][0], pegmove[i][1], pegmove[i][2])
                z = 13 - i
            print("Pegs Remaining: ", z)
            return 1
            pegmove = stack.getElements()
        z = 0
        for i in range(0, peg):
            print(pegmove[i][0], pegmove[i][1], pegmove[i][2])
            z = 13 - i
        print("Pegs Remaining: ", z)
        print("Do you want a better solution? Select 1 or 0")
        better_soln = int(input())
        if (better_soln == 1):
            peg = peg + 1
        else:
            return 1

    t = uploadmymoves(board)
    for i in range(len(t)):
        checkhole = checkhole + 1
        jump(board, t[i][0], 0)
        jump(board, t[i][1], 0)
        jump(board, t[i][2], 1)
        stack.push(t[i])
        if (checkforsolution(board, checkhole)):
            return 1
        else:
            checkhole -= 1
            stack.pop()
            jump(board, t[i][0], 1)
            jump(board, t[i][1], 1)
            jump(board, t[i][2], 0)
    return (0)


print("Initial hole to be empty at:")
set_board = [[-1 for i in range(j + 1)] for j in range(5)]
stack = Stack()
while (1):
    init_hole = int(input())
    #print("Enter the maximum no. of pegs allowed to be left")
    pegs_allowed = 1
    peg = 14 - pegs_allowed
    if (init_hole < 15 and init_hole > -1 and pegs_allowed > 0 and pegs_allowed < 13):
        set_board = findmyrow(set_board, init_hole)
        checkforsolution(set_board, 0)
        break
    else:
        print("Invalid Input. Enter the maximum no. of pegs allowed to be left again")

File: MyCPLHiQ.java
import java.util.Random;
import java.util.Scanner;
import java.nio.charset.Charset;
import java.io.*;

public class MyCPLHiQ {

	static int[] mypegs = new int[15];
	  static int bettersol;
	  static int lines_req=0;
	  static int[][] moves = new int[15][3];
	  static int[][] moves_allowed = new int[100][3];//{{0, 1, 3},{0,2,5},{1, 3, 6},{1, 4, 8},{2, 4, 7},{2, 5, 9},{3, 4, 5},{3, 6, 10},{3, 7, 12},{4, 7, 11},{4, 8, 13},{5, 8, 12},{5, 9, 14},{6, 7, 8},{7, 8, 9},{10, 11, 12},{11, 12, 13},{12, 13, 14}};
	  
	  public static void main(String[] args){

		    int init_hole = 0;
		    int i=0;
		    int better=0;
		    uploadmymoves();
		    Scanner input = new Scanner(System.in);
		    System.out.print("Initial hole to be empty at: ");
		    init_hole = input.nextInt();
		    System.out.print("No of pegs you want to leave at the end of game: ");
		    better = input.nextInt();
		    bettersol = better;
		    for(i=0;i<15;i++) mypegs[i] = 1;
		    mypegs[init_hole] = 0;
		    checkforsolution(0);
		  }
	  
	  public static void uploadmymoves(){
		    try{
		      Scanner sc = new Scanner(new File("mymoves.txt"));
		      int i = 0;
		      String text = "";
		      while (sc.hasNextLine()) {
		        text = sc.nextLine();
		        String[] arr=text.split(" ");
		        for(int j=0;j<arr.length;j++) moves_allowed[lines_req][j]=Integer.parseInt(arr[j]);
		        lines_req++;
		      }
		    }
		    catch(Throwable e){
		      System.out.println(e);
		    }
		  }

	  
	  public static void checkforsolution(int num_moves){

		    int i=0;
		    int j=0;
		    int k=0;
		    int count=0;

		    for(i=0;i<15;i++){
		      if(mypegs[i]==1){
		        count+=1;
		      }
		    }
		    if(count<=bettersol || count==1){
		      System.out.println("No. of Moves: "+num_moves +" \nPegs Remaining: "+count);

		      for(j=0;j<num_moves;j++) System.out.println(moves[j][0] +" " +moves[j][1] +" " +moves[j][2]);
		      System.out.println();
		      System.out.println("Do you want to search for a better solution? (yes/no): ");
		      Scanner scanner = new Scanner(System.in);
		      String resp = scanner.nextLine();
			if(resp.equalsIgnoreCase("no")){
		        File f = new File("output.txt");
		        System.out.println("Thank you!! Solution is in output.txt");
		        try{

		          OutputStream fo = new FileOutputStream(f);
		          for(i=0;i<num_moves;i++){
		            fo.write((moves[i][0]+" "+moves[i][1]+" "+moves[i][2]+"\n").getBytes("utf-8"));
		          }
		          fo.close();

		          System.exit(1);
		        }
		        catch(Throwable e){}
		      }
		      if(bettersol!=1) bettersol--;
		    }

		    for(k=0;k<15;k++){
		      if(mypegs[k]==1) continue;
		      for(i=0;i<15;i++){
		        for(j=0;j<15;j++){
		          if(check_possiblemove(i,j,k)==1){

		            mypegs[i] = 0;
		            mypegs[j] = 0;
		            mypegs[k] = 1;

		            moves[num_moves][0] = i;
		            moves[num_moves][1] = j;
		            moves[num_moves][2] = k;
		            checkforsolution(num_moves+1);

		            mypegs[i] = 1;
		            mypegs[j] = 1;
		            mypegs[k] = 0;
		          }
		        }
		      }
		    }
		  }

	  
	  
	  int findmyrow(int num){
	    int i = 0;
	    int count = 2;
	    if(num==0) return 1;
	    while(i<=num) {
	      i+= count;
	      count++;
	    }
	    return count-1;
	  }

	  public static int check_possiblemove(int from, int over, int to){

	    for(int i=0;i<lines_req;i++){
	      if(mypegs[from]==1 && mypegs[to]==0 && mypegs[over]==1){
	        if(from == moves_allowed[i][0] && over == moves_allowed[i][1] && to==moves_allowed[i][2]) return 1;
	        if(from == moves_allowed[i][2] && over == moves_allowed[i][1] && to==moves_allowed[i][0]) return 1;
	      }
	    }
	    return 0;
	  }

	   

	
}
